### PR TITLE
Remove gpdb specific dtx recovery code for utility mode.

### PR DESCRIPTION
**Pull request by paul-guo-**
Those functionality is handled in checkpoint during recovery in utility
mode also.  The code is not needed anymore.

### CODE REVIEW

File: src/backend/cdb/cdbtm.c [hunk 1]
```diff
@@ -1032,10 +1032,10 @@ tmShmemInit(void)
 	 * We can assign MaxBackends (MaxConnections should be fine also but let's
 	 * be conservative) to max_tm_gxacts on master/standby to tolerate various
 	 * configuration combinations of max_prepared_transactions and
-	 * max_connections. For segments or utility mode, max_tm_gxacts is useless
-	 * so let's set it as zero to save memory.
+	 * max_connections. max_tm_gxacts is used on the coordinator only, and the
+	 * coordinator might be accessed in dispatch mode or utility mode.
 	 */
-	if (Gp_role == GP_ROLE_DISPATCH)
+	if (Gp_role == GP_ROLE_DISPATCH || Gp_role == GP_ROLE_UTILITY)
```
Comment (asimrp) [inline src/backend/cdb/cdbtm.c:1038]: A better way check for coordinator's identity is `IS_QUERY_DISPATCHER()` macro.

**Review by asimrp — COMMENTED** [review]
I was about to approve this PR with a "good riddance" remark, but then the following question propped up in my head: what if crash recovery writes a checkpoint after populating shared memory and the system crashes again, before DTX recovery process gets a chance to process the shared memory.  The shared memory is gone forever and it will never be recreated due to the end-of-recovery checkpoint.  Is this a valid concern?

File: src/backend/cdb/cdbtm.c (hunk 1 above)
Comment (paul-guo-) [reply src/backend/cdb/cdbtm.c]: Agreed and modified.

**Comment by paul-guo-** [comment]
> I was about to approve this PR with a "good riddance" remark, but then the following question propped up in my head: what if crash recovery writes a checkpoint after populating shared memory and the system crashes again, before DTX recovery process gets a chance to process the shared memory. The shared memory is gone forever and it will never be recreated due to the end-of-recovery checkpoint. Is this a valid concern?

@asimrp If I understand the concern correctly, the answer is that if the End-Of-Recovery checkpoint is not flushed into the xlog file due to a new crash just after crash recovery, crash recovery would re-populate the shared memory based on previous checkpoint and 2pc xlog replay after checkpoint.redo.

**Comment by asimrp** [comment]
What happens if the end-of-recovery checkpoint is flushed successfully and then there is a crash?

**Comment by paul-guo-** [comment]
> What happens if the end-of-recovery checkpoint is flushed successfully and then there is a crash?

Then the end-of-recovery checkpoint xlog will include the dtx info that was previously stored in shared memory.

**Review by asimrp — APPROVED** [review]
In that case, good riddance!

:+1:

**Comment by paul-guo-** [comment]
Thanks for reviewing.

**Pull request merged**